File: flows/dashboard-configuration.js
/**
 * Dashboard Configuration Flow
 * Main UI structure for the Smart Home Dashboard
 */
module.exports = function (RED) {
    "use strict";

    // Dashboard configuration
    const dashboard = {
        name: "Smart Home Dashboard",
        version: "1.0.0",
        tabs: {
            dashboard: {
                label: "Dashboard",
                type: "ui_tab",
                order: 1,
                views: {
                    main: {
                        type: "page",
                        name: "Main View",
                        tabs: "row",
                        nodes: []
                    },
                    monitor: {
                        type: "page",
                        name: "Monitor View",
                        nodes: []
                    },
                    devices: {
                        type: "page",
                        name: "Devices",
                        nodes: []
                    },
                    settings: {
                        type: "page",
                        name: "Settings",
                        nodes: []
                    }
                }
            }
        },
        layout: "landscape",
        refresh: 5000,
        theme: "complete"
    };

    // Home Lab Monitor Node
    RED.nodes.addType("home-lab-monitor", {
        label: "Home Lab Monitor",
        type: "ui_gauge",
        common: {
            label: "Node Status",
            format: "%%",
            min: 0,
            max: 100
        }
    });

    // Hubitat Control Node -- carries the device control surface the Devices
    // view renders. Each widget maps onto a Hubitat Maker API command; the UI
    // publishes {"deviceId","command","value"} on input_topic and results come
    // back on output_topic (see hubitat-integration.js controlDevice()).
    RED.nodes.addType("hubitat-control", {
        label: "Hubitat Control",
        type: "hubitat-control",
        common: {
            label: "Device Control",
            inputTopic: "hubitat/device/+/control",
            outputTopic: "hubitat/device/+/command",
            widgets: [
                { command: "on", capability: "switch", widget: "ui_switch" },
                { command: "off", capability: "switch", widget: "ui_switch" },
                { command: "setLevel", capability: "dimmer", widget: "ui_slider", args: ["level", "duration"] },
                { command: "setColor", capability: "color-bulb", widget: "ui_colour_picker", args: ["hue", "saturation", "level"] },
                { command: "setColorTemperature", capability: "color-bulb", widget: "ui_slider", args: ["temperature", "level"] },
                { command: "refresh", capability: "common", widget: "ui_button" }
            ]
        }
    });

    // UniFi Monitor Node
    RED.nodes.addType("unifi-monitor", {
        label: "UniFi Monitor",
        type: "unifi-monitor",
        common: {
            label: "Network Monitor"
        }
    });

    // Kanban Card Node
    RED.nodes.addType("kanban-card", {
        label: "Kanban Card",
        type: "kanban-card",
        common: {
            label: "Task Card"
        }
    });

    return dashboard;
};

File: flows/dashboard_configuration.py
"""Dashboard configuration mirror -- Smart Home Dashboard UI structure.

Python mirror of ``flows/dashboard-configuration.js``: dashboard
name/version, tab and view layout, and node-type registrations.

The ``hubitat-control`` registration also carries the *device control surface*
-- the widgets the Devices view renders and the handler that turns a widget
action into a Hubitat Maker API command --
:func:`device_control_surface` / :func:`handle_device_control`.
"""

DEFAULT_LAYOUT = "landscape"
DEFAULT_REFRESH_MS = 5000
DEFAULT_THEME = "complete"

DASHBOARD_NAME = "Smart Home Dashboard"
DASHBOARD_VERSION = "1.0.0"

#: UI group per device capability -- which control a widget renders.
CAPABILITY_WIDGET_TYPE = {
    "switch": "ui_switch",
    "dimmer": "ui_slider",
    "color-bulb": "ui_colour_picker",
    "common": "ui_button",
}


def _control_widgets():
    """Build the device-control widget descriptors for the Devices view.

    One entry per supported Maker API command, grouped by device capability, so
    the dashboard knows which widget to render and which command name to send.
    """
    from hubitat_integration import HubitatNode

    widgets = []
    for command, spec in sorted(HubitatNode.supportedCommands().items()):
        widgets.append(
            {
                "command": command,
                "capability": spec["capability"],
                "widget": CAPABILITY_WIDGET_TYPE.get(spec["capability"], "ui_button"),
                "args": list(spec["args"]),
                "required": list(spec["required"]),
                "description": spec["description"],
            }
        )
    return widgets


def device_control_surface():
    """Return the dashboard's device-control surface definition.

    This is what the ``hubitat-control`` node publishes to the Devices view:
    the topic the UI publishes control actions on, the topic results come back
    on, and the widget list derived from the supported Maker API commands.
    """
    return {
        "label": "Device Control",
        "type": "hubitat-control",
        "input_topic": "hubitat/device/+/control",
        "output_topic": "hubitat/device/+/command",
        "widgets": _control_widgets(),
    }


def handle_device_control(node, msg):
    """Route handler: map a dashboard control action onto a Hubitat command.

    Args:
        node: A ``HubitatNode`` (or anything exposing ``controlDevice``).
        msg: The Node-RED style message carrying the UI action, e.g.
            ``{"payload": {"deviceId": "101", "command": "setLevel", "value": 50}}``.

    Returns:
        dict: JSON-safe result for the dashboard, from
        :meth:`HubitatNode.controlDevice`.
    """
    result = node.controlDevice(msg)
    return result.to_dict() if hasattr(result, "to_dict") else result


def dashboard_configuration(RED=None):
    """Return the Smart Home Dashboard configuration.

    Args:
        RED: Optional Node-RED runtime object (unused by the pure-Python
            mirror; kept for signature parity with the JS flow).

    Returns:
        dict: Dashboard configuration including the ``home-lab-monitor``,
        ``hubitat-control``, ``unifi-monitor`` and ``kanban-card`` node
        registrations.
    """
    return {
        "name": DASHBOARD_NAME,
        "version": DASHBOARD_VERSION,
        "tabs": {
            "dashboard": {
                "name": "Dashboard",
                "order": 1,
                "views": {
                    "main": {"type": "page", "name": "Main View", "nodes": []},
                    "monitor": {"type": "page", "name": "Monitor View", "nodes": []},
                    "devices": {"type": "page", "name": "Devices", "nodes": []},
                    "settings": {"type": "page", "name": "Settings", "nodes": []},
                },
            }
        },
        "layout": DEFAULT_LAYOUT,
        "refresh": DEFAULT_REFRESH_MS,
        "theme": DEFAULT_THEME,
        "home-lab-monitor": {
            "label": "Node Status",
            "format": "%%",
            "min": 0,
            "max": 100,
        },
        "hubitat-control": device_control_surface(),
        "unifi-monitor": {"label": "Network Monitor"},
        "kanban-card": {"label": "Task Card"},
    }


__all__ = [
    "dashboard_configuration",
    "device_control_surface",
    "handle_device_control",
]
